File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/XmlDeclarationSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class XmlDeclarationSyntax : VisualBasicSyntaxNode
	{
		internal XmlDeclarationOptionSyntax _version;

		internal XmlDeclarationOptionSyntax _encoding;

		internal XmlDeclarationOptionSyntax _standalone;

		public SyntaxToken LessThanQuestionToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)base.Green)._lessThanQuestionToken, base.Position, 0);

		public SyntaxToken XmlKeyword => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)base.Green)._xmlKeyword, GetChildPosition(1), GetChildIndex(1));

		public XmlDeclarationOptionSyntax Version => GetRed(ref _version, 2);

		public XmlDeclarationOptionSyntax Encoding => GetRed(ref _encoding, 3);

		public XmlDeclarationOptionSyntax Standalone => GetRed(ref _standalone, 4);

		public SyntaxToken QuestionGreaterThanToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)base.Green)._questionGreaterThanToken, GetChildPosition(5), GetChildIndex(5));

		internal XmlDeclarationSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal XmlDeclarationSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, PunctuationSyntax lessThanQuestionToken, KeywordSyntax xmlKeyword, XmlDeclarationOptionSyntax version, XmlDeclarationOptionSyntax encoding, XmlDeclarationOptionSyntax standalone, PunctuationSyntax questionGreaterThanToken)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(kind, errors, annotations, lessThanQuestionToken, xmlKeyword, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)version.Green, (encoding != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)encoding.Green) : null, (standalone != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)standalone.Green) : null, questionGreaterThanToken), null, 0)
		{
		}

		public XmlDeclarationSyntax WithLessThanQuestionToken(SyntaxToken lessThanQuestionToken)
		{
			return Update(lessThanQuestionToken, XmlKeyword, Version, Encoding, Standalone, QuestionGreaterThanToken);
		}

		public XmlDeclarationSyntax WithXmlKeyword(SyntaxToken xmlKeyword)
		{
			return Update(LessThanQuestionToken, xmlKeyword, Version, Encoding, Standalone, QuestionGreaterThanToken);
		}

		public XmlDeclarationSyntax WithVersion(XmlDeclarationOptionSyntax version)
		{
			return Update(LessThanQuestionToken, XmlKeyword, version, Encoding, Standalone, QuestionGreaterThanToken);
		}

		public XmlDeclarationSyntax WithEncoding(XmlDeclarationOptionSyntax encoding)
		{
			return Update(LessThanQuestionToken, XmlKeyword, Version, encoding, Standalone, QuestionGreaterThanToken);
		}

		public XmlDeclarationSyntax WithStandalone(XmlDeclarationOptionSyntax standalone)
		{
			return Update(LessThanQuestionToken, XmlKeyword, Version, Encoding, standalone, QuestionGreaterThanToken);
		}

		public XmlDeclarationSyntax WithQuestionGreaterThanToken(SyntaxToken questionGreaterThanToken)
		{
			return Update(LessThanQuestionToken, XmlKeyword, Version, Encoding, Standalone, questionGreaterThanToken);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				2 => _version, 
				3 => _encoding, 
				4 => _standalone, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				2 => Version, 
				3 => Encoding, 
				4 => Standalone, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitXmlDeclaration(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitXmlDeclaration(this);
		}

		public XmlDeclarationSyntax Update(SyntaxToken lessThanQuestionToken, SyntaxToken xmlKeyword, XmlDeclarationOptionSyntax version, XmlDeclarationOptionSyntax encoding, XmlDeclarationOptionSyntax standalone, SyntaxToken questionGreaterThanToken)
		{
			if (lessThanQuestionToken != LessThanQuestionToken || xmlKeyword != XmlKeyword || version != Version || encoding != Encoding || standalone != Standalone || questionGreaterThanToken != QuestionGreaterThanToken)
			{
				XmlDeclarationSyntax xmlDeclarationSyntax = SyntaxFactory.XmlDeclaration(lessThanQuestionToken, xmlKeyword, version, encoding, standalone, questionGreaterThanToken);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(xmlDeclarationSyntax, annotations);
				}
				return xmlDeclarationSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/XmlDeclarationSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class XmlDeclarationSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode
		Friend _version As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax

		Friend _encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax

		Friend _standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax

		Public ReadOnly Property Encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax)(Me._encoding, 3)
			End Get
		End Property

		Public ReadOnly Property LessThanQuestionToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)._lessThanQuestionToken, MyBase.Position, 0)
			End Get
		End Property

		Public ReadOnly Property QuestionGreaterThanToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)._questionGreaterThanToken, Me.GetChildPosition(5), MyBase.GetChildIndex(5))
			End Get
		End Property

		Public ReadOnly Property Standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax)(Me._standalone, 4)
			End Get
		End Property

		Public ReadOnly Property Version As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax)(Me._version, 2)
			End Get
		End Property

		Public ReadOnly Property XmlKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax)._xmlKeyword, Me.GetChildPosition(1), MyBase.GetChildIndex(1))
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal lessThanQuestionToken As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax, ByVal xmlKeyword As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax, ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal questionGreaterThanToken As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(kind, errors, annotations, lessThanQuestionToken, xmlKeyword, DirectCast(version.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax), If(encoding IsNot Nothing, DirectCast(encoding.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax), Nothing), If(standalone IsNot Nothing, DirectCast(standalone.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax), Nothing), questionGreaterThanToken), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitXmlDeclaration(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitXmlDeclaration(Me)
		End Sub

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Select Case i
				Case 2
					syntaxNode = Me._version
					Exit Select
				Case 3
					syntaxNode = Me._encoding
					Exit Select
				Case 4
					syntaxNode = Me._standalone
					Exit Select
				Case Else
					syntaxNode = Nothing
					Exit Select
			End Select
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim version As SyntaxNode
			Select Case i
				Case 2
					version = Me.Version
					Exit Select
				Case 3
					version = Me.Encoding
					Exit Select
				Case 4
					version = Me.Standalone
					Exit Select
				Case Else
					version = Nothing
					Exit Select
			End Select
			Return version
		End Function

		Public Function Update(ByVal lessThanQuestionToken As Microsoft.CodeAnalysis.SyntaxToken, ByVal xmlKeyword As Microsoft.CodeAnalysis.SyntaxToken, ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax, ByVal questionGreaterThanToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Dim xmlDeclarationSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			If (lessThanQuestionToken <> Me.LessThanQuestionToken OrElse xmlKeyword <> Me.XmlKeyword OrElse version <> Me.Version OrElse encoding <> Me.Encoding OrElse standalone <> Me.Standalone OrElse questionGreaterThanToken <> Me.QuestionGreaterThanToken) Then
				Dim xmlDeclarationSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.XmlDeclaration(lessThanQuestionToken, xmlKeyword, version, encoding, standalone, questionGreaterThanToken)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				xmlDeclarationSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, xmlDeclarationSyntax1, xmlDeclarationSyntax1.WithAnnotations(annotations))
			Else
				xmlDeclarationSyntax = Me
			End If
			Return xmlDeclarationSyntax
		End Function

		Public Function WithEncoding(ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(Me.LessThanQuestionToken, Me.XmlKeyword, Me.Version, encoding, Me.Standalone, Me.QuestionGreaterThanToken)
		End Function

		Public Function WithLessThanQuestionToken(ByVal lessThanQuestionToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(lessThanQuestionToken, Me.XmlKeyword, Me.Version, Me.Encoding, Me.Standalone, Me.QuestionGreaterThanToken)
		End Function

		Public Function WithQuestionGreaterThanToken(ByVal questionGreaterThanToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(Me.LessThanQuestionToken, Me.XmlKeyword, Me.Version, Me.Encoding, Me.Standalone, questionGreaterThanToken)
		End Function

		Public Function WithStandalone(ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(Me.LessThanQuestionToken, Me.XmlKeyword, Me.Version, Me.Encoding, standalone, Me.QuestionGreaterThanToken)
		End Function

		Public Function WithVersion(ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationOptionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(Me.LessThanQuestionToken, Me.XmlKeyword, version, Me.Encoding, Me.Standalone, Me.QuestionGreaterThanToken)
		End Function

		Public Function WithXmlKeyword(ByVal xmlKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax
			Return Me.Update(Me.LessThanQuestionToken, xmlKeyword, Me.Version, Me.Encoding, Me.Standalone, Me.QuestionGreaterThanToken)
		End Function
	End Class
End Namespace